### PR TITLE
Implement source-generated DI registration with C# interceptors

### LINKED ISSUE #76: Source-Generated Dependency Injection for AlexaVoxCraft.MediatR
# Source-Generated Dependency Injection for AlexaVoxCraft.MediatR

## Overview

Replace runtime assembly scanning with **compile-time source generation** to produce explicit dependency injection registrations for AlexaVoxCraft handlers and pipeline behaviors. This will improve Lambda cold start performance, enable AOT/trimming scenarios, and provide deterministic, debuggable service registration.

## Current Architecture

AlexaVoxCraft uses a **custom MediatR-style pattern** (not the official MediatR library) with the following key interfaces:

### Request Handlers
```csharp
// Base handler interface with dynamic routing
public interface IRequestHandler
{
    Task<bool> CanHandle(IHandlerInput input, CancellationToken cancellationToken = default);
    Task<SkillResponse> Handle(IHandlerInput input, CancellationToken cancellationToken = default);
}

// Strongly-typed handler for specific request types
public interface IRequestHandler<TRequestType> : IRequestHandler
    where TRequestType : Request
{
}

// Fallback handler that always returns true from CanHandle
public interface IDefaultRequestHandler : IRequestHandler
{
}
```

### Pipeline Infrastructure
```csharp
// Non-generic pipeline behavior (different from MediatR's IPipelineBehavior<,>)
public interface IPipelineBehavior
{
    Task<SkillResponse> Handle(IHandlerInput input, CancellationToken cancellationToken, RequestHandlerDelegate next);
}

// Exception handling
public interface IExceptionHandler
{
    Task<bool> CanHandle(IHandlerInput handlerInput, Exception ex, CancellationToken cancellationToken = default);
    Task<SkillResponse> Handle(IHandlerInput handlerInput, Exception ex, CancellationToken cancellationToken = default);
}

// Request/Response interceptors
public interface IRequestInterceptor { /* ... */ }
public interface IResponseInterceptor { /* ... */ }
```

### Persistence
```csharp
public interface IPersistenceAdapter
{
    // Persistence operations for session/user data
}
```

### Current Registration (Runtime Scanning)

The current `ServiceRegistrar.AddSkillMediatorClasses()` performs runtime reflection to:
1. Scan assemblies for `IRequestHandler<T>` implementations
2. Register `IPipelineBehavior`, `IExceptionHandler`, interceptors, persistence adapters
3. Handle open generics and interface matching
4. Register with appropriate lifetimes (Transient/Singleton based on type)

**Entry Point:**
```csharp
services.AddSkillMediator(configuration, cfg =>
{
    cfg.RegisterServicesFromAssemblyContaining<Program>();
});
```

## Goals

1. **Performance**: Eliminate runtime reflection scanning for faster Lambda cold starts
2. **AOT/Trimming**: Enable Native AOT and IL trimming by removing dynamic type discovery
3. **Determinism**: Produce consistent, ordered, debuggable registrations
4. **Backward Compatibility**: Maintain existing API; add opt-in source generation path
5. **Developer Experience**: Zero-config for common scenarios; attributes for advanced control

### Architecture: Single NuGet Package (Mapperly-Style)

Ship **one NuGet package** containing:
1. **Runtime library**: `AlexaVoxCraft.MediatR.dll` (existing code + new API)
2. **Source generator**: `AlexaVoxCraft.MediatR.Generators.dll` (packaged as analyzer)

Consumers install a single package; the generator runs automatically at compile-time.

### Project Structure

```
/src
  AlexaVoxCraft.MediatR/
    AlexaVoxCraft.MediatR.csproj               # Updated with generator reference
    DI/
      ServiceCollectionExtensions.cs           # Add UseGeneratedRegistrations()
      SkillServiceConfiguration.cs             # Add PreferGenerated flag
      AlexaVoxCraftServiceCollection.stub.cs   # Compile-time stub (IsAvailable = false)
    Attributes/
      AlexaHandlerAttribute.cs                 # [NEW] Lifetime, Exclude control
      PipelineOrderAttribute.cs                # [NEW] Deterministic behavior ordering

  AlexaVoxCraft.MediatR.Generators/            # [NEW] Source generator project
    AlexaVoxCraft.MediatR.Generators.csproj
    Generators/
      AlexaVoxCraftDiGenerator.cs              # Incremental source generator
      SymbolDiscovery.cs                       # Interface matching and symbol walking
      RegistrationEmitter.cs                   # Code generation
      DiagnosticDescriptors.cs                 # Optional analyzer warnings
```

### Packaging Configuration

**Runtime project** (`AlexaVoxCraft.MediatR.csproj`):
```xml
<ItemGroup>
  <ProjectReference Include="..\AlexaVoxCraft.MediatR.Generators\AlexaVoxCraft.MediatR.Generators.csproj"
                    OutputItemType="Analyzer"
                    ReferenceOutputAssembly="false" />
</ItemGroup>
```

**Generator project** (`AlexaVoxCraft.MediatR.Generators.csproj`):
```xml
<PropertyGroup>
  <TargetFramework>netstandard2.0</TargetFramework>
  <LangVersion>latest</LangVersion>
  <IncludeBuildOutput>false</IncludeBuildOutput>
  <EnforceExtendedAnalyzerRules>true</EnforceExtendedAnalyzerRules>
</PropertyGroup>

<ItemGroup>
  <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.*" PrivateAssets="all" />
  <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.*" PrivateAssets="all" />
</ItemGroup>
```

### Updated Configuration Class

```csharp
public class SkillServiceConfiguration
{
    // Existing members...
    internal bool PreferGenerated { get; private set; }

    /// <summary>
    /// Use source-generated DI registrations instead of runtime assembly scanning.
    /// Falls back to scanning if no generated code is available.
    /// </summary>
    public SkillServiceConfiguration UseGeneratedRegistrations()
    {
        PreferGenerated = true;
        return this;
    }

    // Existing: RegisterServicesFromAssemblyContaining<T>() ...
}
```

### Updated Entry Point

```csharp
public static class ServiceCollectionExtensions
{
    public static IServiceCollection AddSkillMediator(
        this IServiceCollection services,
        IConfiguration configuration,
        Action<SkillServiceConfiguration> settingsAction,
        string sectionName = SkillServiceConfiguration.SectionName)
    {
        var serviceConfig = new SkillServiceConfiguration();
        configuration.GetSection(sectionName).Bind(serviceConfig);
        settingsAction.Invoke(serviceConfig);

        // Choose registration path
        if (serviceConfig.PreferGenerated && AlexaVoxCraftServiceCollection.IsAvailable)
        {
            // Use source-generated registrations
            AlexaVoxCraftServiceCollection.AddAlexaVoxCraftGenerated(services);
        }
        else
        {
            // Fallback to runtime scanning (existing behavior)
            if (serviceConfig.AssembliesToRegister.Count == 0)
                throw new ArgumentException("No assemblies found to scan...");

            services.AddRequiredServices(serviceConfig);
            services.AddSkillMediatorClasses(serviceConfig);
        }

        return services;
    }
}
```

### Compile-Time Stub (Non-Generated)

Provides safe fallback when generator doesn't run:

```csharp
// AlexaVoxCraft.MediatR/DI/AlexaVoxCraftServiceCollection.stub.cs
namespace AlexaVoxCraft.MediatR.DI
{
    internal static partial class AlexaVoxCraftServiceCollection
    {
        public const bool IsAvailable = false;

        public static IServiceCollection AddAlexaVoxCraftGenerated(this IServiceCollection services)
            => services;
    }
}
```

The generator will emit the **real** implementation that shadows this with `IsAvailable = true`.

### Handler Control

```csharp
[AttributeUsage(AttributeTargets.Class)]
public sealed class AlexaHandlerAttribute : Attribute
{
    /// <summary>Service lifetime for this handler (default: Transient)</summary>
    public ServiceLifetime Lifetime { get; init; } = ServiceLifetime.Transient;

    /// <summary>Exclude this type from source generation</summary>
    public bool Exclude { get; init; } = false;
}
```

### Pipeline Ordering

```csharp
[AttributeUsage(AttributeTargets.Class)]
public sealed class PipelineOrderAttribute : Attribute
{
    public int Order { get; }
    public PipelineOrderAttribute(int order) => Order = order;
}
```

**Usage:**
```csharp
[PipelineOrder(10)]
public class LoggingBehavior : IPipelineBehavior { /* ... */ }

[PipelineOrder(20)]
public class ValidationBehavior : IPipelineBehavior { /* ... */ }

[AlexaHandler(Lifetime = ServiceLifetime.Scoped)]
public class SessionAwareHandler : IRequestHandler<IntentRequest> { /* ... */ }

[AlexaHandler(Exclude = true)]
public class ExperimentalHandler : IRequestHandler<LaunchRequest> { /* ... */ }
```

### Discovery Phase

The incremental source generator will discover:

1. **Request Handlers** - Types implementing:
   - `IRequestHandler<TRequestType>` where `TRequestType : Request`
   - `IDefaultRequestHandler` (fallback handlers)

2. **Pipeline Behaviors** - Types implementing:
   - `IPipelineBehavior` (non-generic, AlexaVoxCraft-specific)

3. **Exception Handlers** - Types implementing:
   - `IExceptionHandler`

4. **Interceptors** - Types implementing:
   - `IRequestInterceptor`
   - `IResponseInterceptor`

5. **Persistence** - Types implementing:
   - `IPersistenceAdapter`

### Symbol Filtering

```csharp
// Pseudo-code for generator logic
foreach (var type in compilation.GetSymbolsWithName(...))
{
    if (type.HasAttribute("AlexaHandlerAttribute") && attr.Exclude == true)
        continue; // Skip excluded types

    var lifetime = type.GetAttribute("AlexaHandlerAttribute")?.Lifetime
                   ?? ServiceLifetime.Transient;

    var order = type.GetAttribute("PipelineOrderAttribute")?.Order ?? 0;

    if (type.Implements("IRequestHandler<T>"))
        handlers.Add(new HandlerRegistration(type, lifetime));

    if (type.Implements("IPipelineBehavior"))
        behaviors.Add(new BehaviorRegistration(type, lifetime, order));

    // ... similar for other interfaces
}
```

### Generated Output

The generator emits a single file: `AlexaVoxCraft.ServiceCollection.g.cs`

```csharp
// <auto-generated />
#nullable enable
using Microsoft.Extensions.DependencyInjection;

namespace AlexaVoxCraft.MediatR.DI
{
    internal static partial class AlexaVoxCraftServiceCollection
    {
        public const bool IsAvailable = true;

        public static IServiceCollection AddAlexaVoxCraftGenerated(this IServiceCollection services)
        {
            // === Request Handlers ===
            services.AddTransient<IRequestHandler<LaunchRequest>, LaunchRequestHandler>();
            services.AddTransient<IRequestHandler<IntentRequest>, HelpIntentHandler>();
            services.AddScoped<IRequestHandler<IntentRequest>, SessionAwareHandler>();

            // === Default Handlers ===
            services.TryAddTransient<IDefaultRequestHandler, FallbackHandler>();

            // === Pipeline Behaviors (ordered by PipelineOrderAttribute) ===
            services.AddTransient<IPipelineBehavior, LoggingBehavior>();        // Order: 10
            services.AddTransient<IPipelineBehavior, ValidationBehavior>();     // Order: 20
            services.AddTransient<IPipelineBehavior, MetricsBehavior>();        // Order: 30

            // === Exception Handlers ===
            services.AddTransient<IExceptionHandler, GlobalExceptionHandler>();

            // === Interceptors ===
            services.AddTransient<IRequestInterceptor, RequestLoggingInterceptor>();
            services.AddTransient<IResponseInterceptor, ResponseValidationInterceptor>();

            // === Persistence Adapters ===
            services.TryAddSingleton<IPersistenceAdapter, DynamoDbPersistenceAdapter>();

            return services;
        }
    }
}
```

### Registration Logic Details

1. **Request Handlers**:
   - Use `AddTransient()` (allows multiple handlers per request type)
   - Respects `AlexaHandlerAttribute.Lifetime` if present

2. **Default Handlers**:
   - Use `TryAddTransient()` (only one default handler)

3. **Pipeline Behaviors**:
   - Ordered by `PipelineOrderAttribute.Order` (ascending)
   - Ties broken by full type name (deterministic)
   - Use `AddTransient()` to maintain order

4. **Exception Handlers & Interceptors**:
   - Use `AddTransient()` (multiple allowed)

5. **Persistence Adapters**:
   - Use `TryAddSingleton()` (only one adapter)

### Basic Usage (Zero Config)

```csharp
// Program.cs
services.AddSkillMediator(configuration, cfg =>
{
    cfg.UseGeneratedRegistrations();
});
```

Generator automatically discovers all handlers/behaviors in the consumer's compilation.

### Fallback to Scanning

```csharp
// Works if generator doesn't produce output (old projects, unusual build)
services.AddSkillMediator(configuration, cfg =>
{
    cfg.UseGeneratedRegistrations(); // Tries generated first
    cfg.RegisterServicesFromAssemblyContaining<Program>(); // Fallback
});
```

### Advanced: Control Lifetimes

```csharp
[AlexaHandler(Lifetime = ServiceLifetime.Scoped)]
public class UserContextHandler : IRequestHandler<LaunchRequest>
{
    // Will be registered as Scoped instead of default Transient
}
```

### Advanced: Explicit Ordering

```csharp
[PipelineOrder(5)]
public class EarlyLoggingBehavior : IPipelineBehavior { /* ... */ }

[PipelineOrder(100)]
public class LateMetricsBehavior : IPipelineBehavior { /* ... */ }
```

### Unit Tests (Generator)

Create `AlexaVoxCraft.MediatR.Generators.Tests` project with:

1. **Golden File Tests**
   - Input: Small compilation with sample handlers/behaviors
   - Expected: Exact generated code (stored in `Golden/` folder)
   - Validates: Correct interface discovery, registration syntax, ordering

2. **Attribute Tests**
   - `[AlexaHandler(Exclude = true)]` → type omitted from output
   - `[AlexaHandler(Lifetime = Scoped)]` → uses `AddScoped()`
   - `[PipelineOrder(10)]` → correct ordering in output

3. **Edge Cases**
   - No handlers found → empty `AddAlexaVoxCraftGenerated()` body, `IsAvailable = true`
   - Multiple handlers for same request type → all registered
   - Multiple default handlers → only first registered (TryAdd)
   - Open generic types → handled correctly

4. **Compilation Tests**
   - Generated code compiles without errors
   - `IsAvailable` constant is `true` in generated code

### Integration Tests (Runtime)

1. **Registration Parity**
   - Compare services registered by generated code vs. scanning
   - Ensure handler resolution works identically

2. **Lambda Cold Start Benchmark**
   - Measure cold start with scanning vs. generated
   - Target: 30-50% reduction in startup time

3. **Trimming/AOT Validation**
   - Build sample with `PublishTrimmed=true`
   - Build sample with NativeAOT
   - Verify no trimming warnings, application works correctly

## Acceptance Criteria

- ✅ Generator discovers all AlexaVoxCraft-specific interfaces (not MediatR interfaces)
- ✅ Generated registrations produce identical runtime behavior to assembly scanning
- ✅ Pipeline behaviors maintain deterministic ordering via `PipelineOrderAttribute`
- ✅ `[AlexaHandler]` attributes control lifetime and exclusion
- ✅ Fallback to scanning works seamlessly when `IsAvailable = false`
- ✅ Lambda cold start time reduced by measurable amount (benchmark in CI)
- ✅ Works with IL trimming and NativeAOT (no reflection warnings)
- ✅ Golden tests validate exact generated code format
- ✅ Documentation updated (README, CLAUDE.md)
- ✅ Sample projects demonstrate usage

### Phase 1: Generator Infrastructure
- [ ] Create `AlexaVoxCraft.MediatR.Generators` project (netstandard2.0, Roslyn SDK)
- [ ] Implement incremental source generator scaffolding
- [ ] Add symbol discovery for `IRequestHandler<T>`, `IDefaultRequestHandler`
- [ ] Add symbol discovery for `IPipelineBehavior` (non-generic)
- [ ] Add symbol discovery for `IExceptionHandler`, interceptors, persistence
- [ ] Implement attribute parsing (`AlexaHandlerAttribute`, `PipelineOrderAttribute`)
- [ ] Build internal model (HandlerRegistration, BehaviorRegistration, etc.)

### Phase 2: Code Generation
- [ ] Implement registration emitter (generates `AddAlexaVoxCraftGenerated()`)
- [ ] Generate correct registration methods per type (AddTransient, TryAddSingleton, etc.)
- [ ] Implement behavior ordering logic (by Order, then by type name)
- [ ] Handle lifetime overrides from attributes
- [ ] Emit `IsAvailable = true` constant
- [ ] Ensure deterministic output (sorted registrations)

### Phase 3: Runtime Integration
- [ ] Add `AlexaHandlerAttribute` to runtime package
- [ ] Add `PipelineOrderAttribute` to runtime package
- [ ] Add `PreferGenerated` flag to `SkillServiceConfiguration`
- [ ] Implement `UseGeneratedRegistrations()` method
- [ ] Add compile-time stub `AlexaVoxCraftServiceCollection.stub.cs`
- [ ] Update `AddSkillMediator()` to choose generated vs. scanning path

### Phase 4: Packaging
- [ ] Configure generator as Analyzer in runtime csproj (`OutputItemType="Analyzer"`)
- [ ] Verify NuGet package structure (runtime + analyzer embedded)
- [ ] Test package installation in sample project
- [ ] Validate generator runs automatically on consumer builds

### Phase 5: Testing
- [ ] Create `AlexaVoxCraft.MediatR.Generators.Tests` project
- [ ] Implement golden file test infrastructure
- [ ] Add tests for each interface discovery (handlers, behaviors, interceptors, etc.)
- [ ] Add attribute tests (Lifetime, Exclude, Order)
- [ ] Add edge case tests (no handlers, multiple handlers, etc.)
- [ ] Add compilation tests (verify generated code compiles)
- [ ] Add integration tests (compare generated vs. scanning)
- [ ] Add Lambda cold start benchmark (before/after)

### Phase 6: Samples & Documentation
- [ ] Update `Sample.Skill.Function` to use `UseGeneratedRegistrations()`
- [ ] Update `Sample.Apl.Function` to use `UseGeneratedRegistrations()`
- [ ] Add example with `[AlexaHandler]` attributes
- [ ] Add example with `[PipelineOrder]` attributes
- [ ] Update README.md with source generation section
- [ ] Update CLAUDE.md with generator usage instructions
- [ ] Document performance improvements (benchmark results)

### Phase 7: CI/CD
- [ ] Ensure generator tests run in CI
- [ ] Add trimming/AOT validation job to CI
- [ ] Add performance benchmark to CI (track cold start metrics)
- [ ] Verify NuGet package build includes analyzer

### Phase 8: Release
- [ ] Update version in `Directory.Build.props`
- [ ] Add migration guide (scanning → generated)
- [ ] Create release notes highlighting performance gains
- [ ] Tag release (e.g., `v4.1.0`)

## Performance Targets

- **Cold Start Reduction**: 30-50% faster Lambda initialization
- **Compilation Time**: <500ms additional build time for generator
- **Package Size**: <100KB increase (generator dll)
- **Memory**: No increase in runtime memory (same registrations, different path)

### For Existing Projects

**Before:**
```csharp
services.AddSkillMediator(configuration, cfg =>
{
    cfg.RegisterServicesFromAssemblyContaining<Program>();
});
```

**After:**
```csharp
services.AddSkillMediator(configuration, cfg =>
{
    cfg.UseGeneratedRegistrations();
});
```

No other changes required for basic scenarios.

### Opt-Out (Keep Scanning)

Simply don't call `UseGeneratedRegistrations()`:
```csharp
services.AddSkillMediator(configuration, cfg =>
{
    cfg.RegisterServicesFromAssemblyContaining<Program>(); // Still works
});
```

## Future Enhancements

- Analyzer diagnostics (warn about missing handlers, duplicate default handlers, etc.)
- Support for handler discovery across referenced projects
- Metrics dashboard showing registration performance
- Automatic ordering suggestions for pipeline behaviors

## References

- Mapperly source generation approach: [Mapperly GitHub](https://github.com/riok/mapperly)
- Roslyn incremental generators: [Microsoft Docs](https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview)
- AlexaVoxCraft current registration: `src/AlexaVoxCraft.MediatR/Registration/ServiceRegistrar.cs`

### PR DESCRIPTION
**Pull request by ncipollina**
## Summary

Fixes #76 - Implements compile-time source-generated dependency injection registration using C# interceptors, replacing runtime reflection-based assembly scanning. This provides zero-overhead service registration with full compile-time type safety.

### C# Interceptor-Based Code Generation
- ✅ Intercepts `AddSkillMediator()` calls at compile time
- ✅ Generates optimized registration code with no runtime reflection
- ✅ Automatic discovery of handlers, behaviors, interceptors, and exception handlers
- ✅ Preserves exact method signature including optional parameters

### Automatic MSBuild Integration
- ✅ MSBuild props files automatically configure `InterceptorsNamespaces` property
- ✅ Works for both direct package consumers (`build/`) and transitive references (`buildTransitive/`)
- ✅ No manual project configuration required
- ✅ Compatible with local project references and NuGet packages

### Generator Requirements & Safety
- ✅ Requires C# 11 or higher (validated at compile time)
- ✅ Requires .NET 8.0 or higher for interceptor support
- ✅ Can be disabled via `EnableMediatRGeneratorInterceptor=false` MSBuild property
- ✅ Only intercepts public `AddSkillMediator` method with `IConfiguration` parameter

### Generated Code Features
- ✅ Registers all request handlers with proper lifetime management
- ✅ Orders pipeline behaviors by `Order` property
- ✅ Registers exception handlers and request/response interceptors
- ✅ Registers persistence adapters (TryAddSingleton for single instance)
- ✅ Properly configures `IOptions<SkillServiceConfiguration>` for DI injection
- ✅ Includes diagnostic warnings for multiple default handlers or persistence adapters

### Generator Architecture
```csharp
// User code - unchanged
services.AddSkillMediator(configuration, cfg => 
    cfg.AssembliesToRegister.Add(typeof(Program).Assembly));

// Generated interceptor (simplified)
[InterceptsLocation(version: 1, data: "hash")]
internal static IServiceCollection AddSkillMediator(...)
{
    var cfg = new SkillServiceConfiguration();
    configuration.GetSection(sectionName).Bind(cfg);
    settingsAction?.Invoke(cfg);
    
    services.Configure<SkillServiceConfiguration>(opt => { ... });
    ServiceRegistrar.AddRequiredServices(services, cfg);
    
    // Generated registrations for all discovered types
    services.AddTransient<IRequestHandler<LaunchRequest>, LaunchRequestHandler>();
    services.AddScoped<IPipelineBehavior, LoggingBehavior>();
    // ... etc
    
    return services;
}
```

### MSBuild Integration
```xml
<!-- Automatically included via NuGet package -->
<PropertyGroup>
  <InterceptorsNamespaces>$(InterceptorsNamespaces);AlexaVoxCraft.Generated</InterceptorsNamespaces>
</PropertyGroup>
```

### Symbol Discovery
- Scans all types in compilation for marker interfaces
- Discovers: `IRequestHandler`, `IDefaultRequestHandler`, `IPipelineBehavior`, `IExceptionHandler`, `IRequestInterceptor`, `IResponseInterceptor`, `IPersistenceAdapter`
- Respects `[ServiceLifetime]` attribute for custom lifetimes
- Orders behaviors by `[Order]` attribute
- Validates single default handler and single persistence adapter

### New Files
- `src/AlexaVoxCraft.MediatR.Generators/Constants.cs` - Generator constants
- `src/AlexaVoxCraft.MediatR/build/AlexaVoxCraft.MediatR.props` - Direct consumer props
- `src/AlexaVoxCraft.MediatR/buildTransitive/AlexaVoxCraft.MediatR.props` - Transitive props

### Modified Files
- `AlexaVoxCraftDiGenerator.cs` - Simplified generator with C# 11+ check and interceptor validation
- `InterceptorEmitter.cs` - Added `services.Configure<SkillServiceConfiguration>()` block
- `DiagnosticDescriptors.cs` - Removed unnecessary diagnostics
- `AlexaVoxCraft.MediatR.csproj` - Simplified analyzer packaging, added props files

## Breaking Changes

None - fully backward compatible. The original `ServiceCollectionExtensions.AddSkillMediator()` method remains in place and is intercepted transparently.

## Performance Benefits

- **Zero runtime reflection** - All type discovery happens at compile time
- **Faster startup** - No assembly scanning on application startup  
- **Smaller IL** - Direct method calls instead of reflection invocations
- **AOT-friendly** - Compatible with Native AOT compilation
- **Compile-time validation** - Catches registration errors at build time

## Migration Notes

For package consumers:
- No code changes required
- Interceptor automatically replaces runtime registration
- Remove any manual `InterceptorsNamespaces` configuration (now automatic)
- Set `EnableMediatRGeneratorInterceptor=false` to disable if needed

## Testing

- ✅ Solution builds successfully across all target frameworks (net8.0, net9.0, net10.0)
- ✅ Sample.Generated.Function demonstrates working interceptor
- ✅ All 491 existing tests pass
- ✅ Generated code includes all service registrations
- ✅ Configuration properly registered for IOptions<T> injection

## Future Enhancements

Potential improvements for future iterations:
- Source link support for generated files
- Additional diagnostics for common configuration errors
- Support for custom registration conventions
- Performance benchmarks vs. reflection approach

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Co-Authored-By: Claude <noreply@anthropic.com>

### CODE REVIEW

**Comment by ncipollina** [comment]
@codex review

File: src/AlexaVoxCraft.MediatR.Generators/Models/RegistrationModel.cs
Comment (j-d-ha) [inline src/AlexaVoxCraft.MediatR.Generators/Models/RegistrationModel.cs:1]: If this class is used to pass data between parts of your generator, you probably want to use a value type as it provides value equality semantics. This is used for caching and with reference types, caching wont work correctly and the whole pipeline will run pretty much every time you hit save or compile.

For my stuff, I added a custom array type that implemented value based equality and for all my objects, I used readonly record structs.

File: src/AlexaVoxCraft.MediatR.Generators/Generators/AlexaVoxCraftDiGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            .CreateSyntaxProvider(
+                static (node, _) => IsAddSkillMediatorInvocation(node),
+                static (ctx, _) => GetCallSiteLocation(ctx))
+            .Where(static loc => loc.HasValue)
+            .Select(static (loc, _) => loc!.Value);
+
+        var allTypes = context.SyntaxProvider
+            .CreateSyntaxProvider(
+                static (node, _) => node is TypeDeclarationSyntax,
+                static (ctx, _) => ctx.SemanticModel.GetDeclaredSymbol((TypeDeclarationSyntax)ctx.Node))
```
Comment (j-d-ha) [inline src/AlexaVoxCraft.MediatR.Generators/Generators/AlexaVoxCraftDiGenerator.cs]: You probably want to extract the info from the type here rather than pulling out the whole type. Due to the location data on the type, you can have some bad caching issues.

File: src/AlexaVoxCraft.MediatR.Generators/Generators/InterceptorEmitter.cs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+namespace AlexaVoxCraft.MediatR.Generators.Generators;
+
+internal static class InterceptorEmitter
+{
+    public static string EmitInterceptors(ImmutableArray<InterceptorLocation> locations, RegistrationModel model)
+    {
+        var sb = new StringBuilder();
+
+        sb.AppendLine("// <auto-generated />");
```
Comment (j-d-ha) [inline src/AlexaVoxCraft.MediatR.Generators/Generators/InterceptorEmitter.cs:14]: If you find you dont like using a string builder, I had a lot of luck with this package: https://github.com/scriban/scriban

File: src/AlexaVoxCraft.MediatR.Generators/Models/RegistrationModel.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    public HandlerRegistration(INamedTypeSymbol type, INamedTypeSymbol? requestType, int lifetime, int order, Location location)
+    {
+        Type = type;
+        RequestType = requestType;
+        Lifetime = lifetime;
+        Order = order;
+        Location = location;
+    }
+
+    public INamedTypeSymbol Type { get; }
```
Comment (j-d-ha) [inline src/AlexaVoxCraft.MediatR.Generators/Models/RegistrationModel.cs]: I would also make an abstraction over ISymbol which is a value type that also just has the data you need.

**Review by j-d-ha — COMMENTED** [review]
👍 Looking really cool! Left some suggestions on performance optimization.

File: src/AlexaVoxCraft.MediatR.Generators/Generators/InterceptorEmitter.cs (hunk 2 above)
Comment (ncipollina) [reply src/AlexaVoxCraft.MediatR.Generators/Generators/InterceptorEmitter.cs:14]: I'm fine w/ StringBuilder for now.

File: src/AlexaVoxCraft.MediatR.Generators/Models/RegistrationModel.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    TypeInfo? RequestType,
+    int Lifetime,
+    int Order
+) : IEquatable<HandlerRegistration>;
+
+internal readonly record struct BehaviorRegistration(
+    TypeInfo Type,
+    int Lifetime,
+    int Order
+) : IEquatable<BehaviorRegistration>;
```
Comment (j-d-ha) [inline src/AlexaVoxCraft.MediatR.Generators/Models/RegistrationModel.cs]: Rider is telling me that IEquatable is implicitly implemented for these types.

File: src/AlexaVoxCraft.MediatR.Generators/Generators/AlexaVoxCraftDiGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        // Treat default/empty as "unavailable"
+        if (il is null)
+            return null;
+
+        return new InterceptorLocation(il.Data);
+    }
+
+}
+
+internal readonly struct InterceptorLocation
```
Comment (j-d-ha) [inline src/AlexaVoxCraft.MediatR.Generators/Generators/AlexaVoxCraftDiGenerator.cs]: Massive nit on my part, but why not use a primary constructor here and also make it a record struct?

File: src/AlexaVoxCraft.MediatR.Generators/Generators/SymbolDiscovery.cs [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    private const string AlexaHandlerAttributeName = "AlexaVoxCraft.MediatR.Annotations.AlexaHandlerAttribute";
+    private const string IRequestHandlerName = "AlexaVoxCraft.MediatR.IRequestHandler";
+    private const string IDefaultRequestHandlerName = "AlexaVoxCraft.MediatR.IDefaultRequestHandler";
+    private const string IPipelineBehaviorName = "AlexaVoxCraft.MediatR.Pipeline.IPipelineBehavior";
+    private const string IExceptionHandlerName = "AlexaVoxCraft.MediatR.Pipeline.IExceptionHandler";
+    private const string IRequestInterceptorName = "AlexaVoxCraft.MediatR.Pipeline.IRequestInterceptor";
+    private const string IResponseInterceptorName = "AlexaVoxCraft.MediatR.Pipeline.IResponseInterceptor";
+    private const string IPersistenceAdapterName = "AlexaVoxCraft.MediatR.Attributes.Persistence.IPersistenceAdapter";
+
+    public static (RegistrationModel model, ImmutableArray<DiagnosticInfo> diagnostics) BuildModel(ImmutableArray<INamedTypeSymbol> symbols)
```
Comment (j-d-ha) [inline src/AlexaVoxCraft.MediatR.Generators/Generators/SymbolDiscovery.cs]: Might want to use EquatableArray here:
```suggestion
    public static (RegistrationModel model, EquatableArray<DiagnosticInfo> diagnostics) BuildModel(ImmutableArray<INamedTypeSymbol> symbols)
```

Comment (j-d-ha) [reply src/AlexaVoxCraft.MediatR.Generators/Generators/SymbolDiscovery.cs]: But that would only be needed if you call this before `RegisterSourceOutput` (which you should also do 😄).

**Review by j-d-ha — COMMENTED** [review]
A few more suggestions, but it looks great!

File: src/AlexaVoxCraft.MediatR.Generators/Generators/AlexaVoxCraftDiGenerator.cs (hunk 5 above)
Comment (ncipollina) [reply src/AlexaVoxCraft.MediatR.Generators/Generators/AlexaVoxCraftDiGenerator.cs]: Good suggestion.

File: src/AlexaVoxCraft.MediatR.Generators/Models/RegistrationModel.cs (hunk 4 above)
Comment (ncipollina) [reply src/AlexaVoxCraft.MediatR.Generators/Models/RegistrationModel.cs]: ➕ for Rider!

**Comment by ncipollina** [comment]
@codex review

**Comment by ncipollina** [comment]
@codex review
